src/GitServer/GitServer.ts

```diff
@@ -0,0 +1,298 @@
+        return true
+      }
+
+      const service = (<string>params.service).replace(/^git-/, '')
+      if (services.indexOf(service) < 0) {
+        res.statusCode = 405
+        res.end('service not available')
+        return true
+      }
+
+      const repoName = this.parseGitName(m[1])
+
+      // check if the repo is authenticated
+      const type = this.getType(service)
+      const headers = req.headers
+      const { username, password } = {username: 'foo', password: 's'}
+      let authenticated: boolean = false
+
+      if (username && password) {
+        authenticated = this.authenticate(type, repoName, username, password, headers)
+      }
+
+      if (!authenticated) {
+        res.setHeader("Content-Type", 'text/plain')
+        res.setHeader("WWW-Authenticate", 'Basic realm="authorization needed"')
+        res.writeHead(401)
+        res.end('Unauthorised to view this repo')
+        return true
+      }
+
+      this.infoResponse(repo, service, req, res)
+      return true
+    }
+    const requestPackHandler = (req: http.IncomingMessage, res: http.ServerResponse): boolean => {
+      if (req.method !== 'POST') return false
+      const m = req.url!.match(/\/(.+)\/git-(.+)/)
+      if (!m) return false
+      if (/\.\./.test(m[1])) return false
+
+      const repo = m[1]
+      const service = m[2]
+
+      if (services.indexOf(service) < 0) {
+        res.statusCode = 405
+        res.end('service not available')
+        return true
+      }
+
+      res.setHeader('content-type', 'application/x-git-' + service + '-result')
+      this.noCache(res)
+
+      const repoDir = Path.join(this.repoDir, repo)
+
+      const fileSystem = this.vaultStore.getVault(repo)?.efs
+      if (fileSystem) {
+        const dup = new HttpDuplex(
+          req,
+          res,
+          repo,
+          service,
+          repoDir,
+          fileSystem
+        )
+      }
+
+      return true
+    }
+    const notSupportedHandler = (req: http.IncomingMessage, res: http.ServerResponse): boolean => {
+      if (req.method !== 'GET' && req.method !== 'POST') {
+        res.statusCode = 405
+        res.end('method not supported')
+        return true
+      } else {
+        return false
+      }
+    }
+    const notFoundHandler = (req: http.IncomingMessage, res: http.ServerResponse): boolean => {
+      res.statusCode = 404
+      res.end('not found')
+      return true
+    }
+
+    const handlers = [
+      infoHandler,
+      requestPackHandler,
+      notSupportedHandler,
+      notFoundHandler
+    ]
+
+    res.setHeader('connection', 'close')
+
+    for (const handler of handlers) {
+      const fulfilled = handler(req,res)
+      if (fulfilled) {
+        break
+      }
+    }
+  }
+
+  /**
+   * starts a git server on the given port
+   */
+  listen(port?: number): AddressInfo {
+    this.server = http.createServer((req, res) => {
+      this.handle(req, res)
+    })
+
+    this.server.listen(port)
```

Comment:
If the port is optional , does this mean this finds a random port? I'd prefer for this to be explicit with `port: number = 0`.

Reply:
yeah this was originally just implicitly null, but we can make it explicitly random with port of `0`

Reply:
Eventually, this will be listening on a pre-made secure connection and we won't need to provide a port.